File: day2/main.py
reports = []

with open('./inputs/input2.txt', 'r') as file: 
    for line in file:
        reports.append(list(map(int, line.split())))

def check_sorted(report):
    return sorted(report) == report or sorted(report, reverse=True) == report

def check_difference(report):
    return all(0 < abs(report[i] - report[i+1]) <= 3 for i in range(len(report) - 1))

def is_strictly_safe(report):
    return check_difference(report) and check_sorted(report)

def count_safe_reports(reports):
    safe_or_not = [is_strictly_safe(report) for report in reports]
    return sum(safe_or_not)
       
print(count_safe_reports(reports))

################## Part 2 ###################


def check_almost_sorted(report):
    def safe_if_almost_sorted(report, sorted_report):
        for i in range(len(report)):
            if report[i] != sorted_report[i]:
                report_no_x = report.copy()
                report_no_x.pop(i)
                report_no_y = report.copy()
                report_no_y.pop(i)
                if is_strictly_safe(report_no_x) or is_strictly_safe(report_no_y):
                    return True
        return False

    strictly_sorted = check_sorted(report)
    if strictly_sorted == False:
        return safe_if_almost_sorted(report, sorted(report)) or safe_if_almost_sorted(report, sorted(report, reverse=True))
    return False

def check_almost_safe_differences(report):
    def offending_indices(report):
        offences = []
        for i in range(len(report)-1):
            if not(0 < abs(report[i] - report[i+1]) <= 3):
                offences.append((i, i+1))
        return offences
    
    offences = offending_indices(report)
    if len(offences) == 1:
        x, y = offences[0]
        report_no_x = report.copy()
        report_no_x.pop(x)
        report_no_y = report.copy()
        report_no_y.pop(y)
        return is_strictly_safe(report_no_x) or is_strictly_safe(report_no_y)
    return False

def count_kinda_safe_reports(reports):
    safe_or_not = []
    for report in reports:
        kinda_safe = check_almost_sorted(report) or check_almost_safe_differences(report)
        safe_or_not.append(kinda_safe)
    return sum(safe_or_not)


########### Naive solution ##################

def brute_force(reports):
    kinda_safe = 0
    for report in reports:
        if not is_strictly_safe(report):
            for i, _ in enumerate(report):
                report_no_level = report.copy()
                report_no_level.pop(i)
                if is_strictly_safe(report_no_level):
                    kinda_safe += 1
                    break
    return kinda_safe


print(count_safe_reports(reports) + count_kinda_safe_reports(reports))
print(count_safe_reports(reports) + brute_force(reports))

File: day2/main.go
package main

import (
	"bufio"
	"os"
	"sort"
	"strconv"
	"strings"
	"math"
	"fmt"
)

func readInput() [][]int {
	var reports [][]int

	file, _ := os.Open("./inputs/input2.txt")
	
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		strNumbers := strings.Fields(line)
		var intNumbers []int
		for _, str := range strNumbers {
			num, _ := strconv.Atoi(str)
			intNumbers = append(intNumbers, num)
		}
		reports = append(reports, intNumbers)
	}

	return reports
}

func checkSorted(report []int) bool {
	isAsc := sort.IntsAreSorted(report)

	isDesc := true

	for i := 1; i < len(report); i++ {
		if report[i] > report[i-1]{
			isDesc = false
		}
	}

	return isAsc || isDesc
} 

func checkDifference(report []int) bool {
	valid := true

	for i := 0; i < len(report) - 1; i++ {
		if  1 > math.Abs(float64(report[i]-report[i+1])) || math.Abs(float64(report[i]-report[i+1])) > 3 {
			valid = false
			break
		} 
	}
	return valid
}

func isStrictlySafe(report []int) bool {
	return checkSorted(report) && checkDifference(report)
}

func countSafeReports(reports [][]int) int {
	sum := 0
	for i := 0; i < len(reports); i++ {
		if isStrictlySafe(reports[i]){
			sum += 1
		}
	}
	return sum
}

func removeAtIndex(arr []int, index int) []int {
	newArr := make([]int, len(arr)-1)
	copy(newArr, arr[:index])
	copy(newArr[index:], arr[index+1:])
	return newArr
}


func countCanBeSafeReports(reports [][]int) int {
	sum := 0

	for i := 0; i < len(reports); i++ {
		if !isStrictlySafe(reports[i]){
			for j:=0; j < len(reports[i]); j++ {
				report_no_level := removeAtIndex(reports[i], j)
				if isStrictlySafe(report_no_level){
					sum += 1
					break
				}
			}
		}
	}
	return sum
}


func main() {
	reports := readInput()
	fmt.Println(countSafeReports(reports))
	fmt.Println(countSafeReports(reports)+ countCanBeSafeReports(reports))
}